lib/submission_runners/crystal.rb

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+require 'submission_runners/base'
+require 'fileutils'
+require_relative '../../lib/docker/helpers'
+require_relative '../support/temp_file_block'
+
+module SubmissionRunners
+  class Crystal < Base
+    attr_reader :entry
+    include Docker::Helpers
+
+    def self.image
+      'crystallang/crystal:0.22.0'
+    end
+
+    def build
+      FileUtils.chmod 0777, submission_dir
+
+      @entry = EntryFile(source_file)
+      CrystalBuildResponse.new \
+        docker_run('crystal', 'build', '-o', entry.basename, entry.filename, chdir: submission_dir)
+    end
+
+    def run
+      docker_run("./#{entry.basename}", chdir: submission_dir, in: input_buffer)
+    end
+
+    class CrystalBuildResponse
+      delegate_missing_to :@result_object
+      def initialize(result_object)
+        @result_object = result_object
+      end
+
+      def out
```

Comment:
There is not content meant for STDOUT when the exit status is not zero.  See https://github.com/crystal-lang/crystal/issues/4451 .  There is no content for `out`.